lib/core/src/recover/recoverer.rs

```diff
@@ -66,25 +93,40 @@ impl Recoverer {
             anyhow!("Got {claim_txs_len} send claim transactions, expected {claim_tx_ids_len}")
         );
 
-        let claim_txs_by_swap_id: HashMap<&String, lwk_wollet::elements::Transaction> =
-            claim_tx_ids_by_swap_id.into_keys().zip(claim_txs).collect();
+        let claim_txs_by_swap_id: HashMap<String, lwk_wollet::elements::Transaction> =
+            failed_cooperative.into_keys().zip(claim_txs).collect();
 
-        let mut preimages = HashMap::new();
         for (swap_id, claim_tx) in claim_txs_by_swap_id {
-            match Self::get_send_swap_preimage_from_claim_tx(swap_id, &claim_tx) {
-                Ok(preimage) => {
-                    preimages.insert(swap_id.to_string(), preimage);
-                }
+            let Some(recovered_data) = recovered_send_data.get_mut(&swap_id) else {
+                continue;
+            };
+
+            match Self::get_send_swap_preimage_from_claim_tx(&swap_id, &claim_tx) {
+                Ok(preimage) => recovered_data.preimage = Some(preimage),
                 Err(e) => {
-                    debug!(
-                        "Couldn't get swap preimage from claim tx {} for swap {swap_id}: {e} - \
-                        could be a cooperative claim tx",
+                    error!(
+                        "Couldn't get non-cooperative swap preimage from claim tx {} for swap {swap_id}: {e}",
                         claim_tx.txid()
                     );
+                    // Keep only claim tx for which there is a recovered or synced preimage
+                    recovered_data.claim_tx_id = None;
                 }
             }
         }
-        Ok(preimages)
+
+        Ok(())
+    }
+
+    async fn recover_preimages(
+        &self,
+        mut recovered_send_data: HashMap<String, &mut RecoveredOnchainDataSend>,
+    ) -> Result<()> {
+        // Recover the preimages by querying the swapper, only if there is a claim_tx_id
+        let failed_cooperative = self.recover_cooperative_preimages(&mut recovered_send_data);
+
+        // For those which failed, recover the preimages by querying onchain (non-cooperative case)
+        self.recover_non_cooperative_preimages(&mut recovered_send_data, failed_cooperative)
+            .await
```

Comment:
We need to make sure the preimages are validated before exiting this function (and assigning them to the swaps).